crates/prover/src/runner.rs

```diff
@@ -202,85 +299,139 @@ where
             None => get_initial_slot_height::<Da::Spec>(&self.sequencer_client).await,
         };
 
-        let mut l2_height = self.start_height;
-
         let prover_config = self.prover_config.clone().unwrap();
 
-        let pg_client = match prover_config.db_config {
+        let pg_client = match prover_config.clone().db_config {
             Some(db_config) => {
                 info!("Connecting to postgres");
                 Some(PostgresConnector::new(db_config.clone()).await)
             }
             None => None,
         };
 
-        loop {
-            let da_service = &self.da_service;
-
-            let exponential_backoff = ExponentialBackoffBuilder::new()
-                .with_initial_interval(Duration::from_secs(1))
-                .with_max_elapsed_time(Some(Duration::from_secs(5 * 60)))
-                .build();
-            let last_finalized_height = retry_backoff(exponential_backoff.clone(), || async {
-                da_service
-                    .get_last_finalized_block_header()
-                    .await
-                    .map_err(backoff::Error::transient)
-            })
-            .await?
-            .height();
+        // Save the genesis state root to ledger db
+        if self.start_l2_height == 1 {
+            self.ledger_db
+                .set_l2_state_root(0, &self.state_root.clone())?;
+        }
 
-            if l1_height > last_finalized_height {
-                sleep(Duration::from_secs(1)).await;
-                continue;
+        let (l1_tx, mut l1_rx) = mpsc::channel(1);
+        let da_service = self.da_service.clone();
+        let l1_block_cache = self.l1_block_cache.clone();
+        let l1_handle = tokio::spawn(async move {
+            l1_sync(start_l1_height, da_service, l1_tx, l1_block_cache).await;
+        });
+        tokio::pin!(l1_handle);
+
+        let (l2_tx, mut l2_rx) = mpsc::channel(1);
+        let start_l2_height = self.start_l2_height;
+        let sequencer_client = self.sequencer_client.clone();
+        let sync_blocks_count = self.sync_blocks_count;
+        let l2_handle = tokio::spawn(async move {
+            sync_l2::<Da>(start_l2_height, sequencer_client, l2_tx, sync_blocks_count).await;
+        });
+        tokio::pin!(l2_handle);
+        let da_service = self.da_service.clone();
+        let l1_block_cache = self.l1_block_cache.clone();
+        let mut pending_l1_blocks: VecDeque<<Da as DaService>::FilteredBlock> =
+            VecDeque::<Da::FilteredBlock>::new();
+        let pending_l1 = &mut pending_l1_blocks;
+        loop {
+            select! {
+                _ = &mut l1_handle => {panic!("l1 sync handle exited unexpectedly");},
+                _ = &mut l2_handle => {panic!("l2 sync handle exited unexpectedly");},
+                Some(l1_block) = l1_rx.recv() => {
+                    pending_l1.push_back(l1_block);
+                 },
+                _ = sleep(Duration::from_secs(1)) => {
```

Comment:
This would instantiate a sleep future every iteration of the loop. Please use `tokio::time::interval` by instantiating it outside of the select. Like what the sequencer does.

Reply:
Not sure why we need interval here, if what we need is the inner function to run every second (except when it detects commitments and tries to prove them ofc) then this works fine.

In essence is this any different than the receiver? It awaits that branch for a second just like receiver does until it receives a message, am I missing something here?

Reply:
You're instantiating a sleep instance on every call to select which could potentially lead to misbehavior. It is always recommended to instantiate the future outside the select and `.await` it inside the select. That way you guarantee it will not misbehave. 

However, when thinking about it some more, here what you're doing is that you're always waiting for 1 second if the `select` randomly selects this match arm to execute first. Which might not be necessary. Using the interval is much more suited to say that i am doing to process pending L1 blocks every 1 second.

Reply:
Okay, just pushed a commit, hope I got it right